Refactor: the method "cleanCharRegions" in Class "CharacterSegmenter"

henryyncn opened the pull request:
Cleaning characters that are too small in "charRegions", just remove them from pipeline_data, rather than fill the charRegions with black color.
	modified:   src/openalpr/ocr/segmentation/charactersegmenter.cpp
	modified:   src/openalpr/ocr/segmentation/charactersegmenter.h

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openalpr/openalpr?pullRequest=619) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openalpr/openalpr?pullRequest=619) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openalpr/openalpr?pullRequest=619) it.</sub>

matthill commented:
Hey Henry,

I ran it through a benchmark of "hard" US plates and it dropped the accuracy by a few points.  Not sure why exactly.  On European plates, it came out about the same.

Before:

    Detected:           303  / 416 : 72.84%
    Top N:              235  / 416 : 56.49%
    Top Pattern:        165  / 416 : 39.66%
    Top 1:              165  / 416 : 39.66%
    State Match:        0    / 195 : 0.00%
    False Positives:    29   / 416 : 6.97%


After:

    Detected:           294  / 416 : 70.67%
    Top N:              224  / 416 : 53.85%
    Top Pattern:        161  / 416 : 38.70%
    Top 1:              161  / 416 : 38.70%
    State Match:        0    / 190 : 0.00%
    False Positives:    26   / 416 : 6.25%

francoamato commented:
UNSUBSCRIBE

On 30 January 2018 at 09:35, Matthew Hill <notifications@github.com> wrote:

> Hey Henry,
>
> I ran it through a benchmark of "hard" US plates and it dropped the
> accuracy by a few points. Not sure why exactly. On European plates, it came
> out about the same.
>
> Before:
>
> Detected:           303  / 416 : 72.84%
> Top N:              235  / 416 : 56.49%
> Top Pattern:        165  / 416 : 39.66%
> Top 1:              165  / 416 : 39.66%
> State Match:        0    / 195 : 0.00%
> False Positives:    29   / 416 : 6.97%
>
> After:
>
> Detected:           294  / 416 : 70.67%
> Top N:              224  / 416 : 53.85%
> Top Pattern:        161  / 416 : 38.70%
> Top 1:              161  / 416 : 38.70%
> State Match:        0    / 190 : 0.00%
> False Positives:    26   / 416 : 6.25%
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openalpr/openalpr/pull/619#issuecomment-361611499>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKEFK-mocXSDnFKipdqrk61jt69349Fpks5tPyi0gaJpZM4Rrmr->
> .
>



-- 
*Franco Amato*
Il caffè è un piacere, se non è buono che piacere è...

henryyncn commented:
Hello matthill,

First of all, thank you for this nice project, I'm developing a Chinese license plate recognition system based on openalpr. 

In my project, I changed the OCR module. I utilized a Convolutional Neural Network for character classification, instead of Tesseract. The original operation (fill the charRegions with balck color, and Tesseract recognizes it  as space character) leads to bugs, because my OCR engine does not support space character, LOL. I think it not necessary to recognize space character in license plate recognition system, so I refactored the method `cleanCharRegions`.

I didn't test US plates or European plates, so I am not sure the problem. As far as I am concerned, the reasons are as follows.
The `pipeline_data` contains 3 `thresholds`, that share the same `charRegions`. As result, when it ran the first loop in `for (unsigned int i = 0; i < thresholds.size(); i++)`, small regions were removed, in next loops, the `charRegions` were modified in last loop, which may cause the deletion of some "thin" characters like `"1"`.

I'm still working on it to optimize the performance of character segmentation, if there is any progress, I'm willing to share my mothod.
Thanks!

matthill commented:
Thanks, Henry.

I understand.  Yes, I think this is a good contribution.  In order to merge it as the default, it needs to improve overall accuracy.  In this case, it seems to improve accuracy in some cases, so it's probably best to enable it with a config option.  That way, you can turn it on/off for cases where it will help.
